apps/wallet/src/core/compatibility.core.ts

```diff
@@ -0,0 +1,248 @@
+ */
+async function fetchStationApiVersion(agent: HttpAgent, stationId: Principal): Promise<string> {
+  const encoder = new TextEncoder();
+  const versionPath: ArrayBuffer[] = [
+    encoder.encode('canister'),
+    stationId.toUint8Array(),
+    encoder.encode('metadata'),
+    encoder.encode('app:version'),
+  ];
+
+  const state = await agent.readState(stationId, {
+    paths: [versionPath],
+  });
+
+  const certificate = await Certificate.create({
+    canisterId: stationId,
+    certificate: state.certificate,
+    rootKey: agent.rootKey,
+  });
+
+  const version = certificate.lookup(versionPath);
+
+  if (!version) {
+    throw new Error('Version not found');
+  }
+
+  const decoder = new TextDecoder();
+  const decodedVersion = decoder.decode(version);
+
+  if (!isSemanticVersion(decodedVersion)) {
+    throw new Error(
+      'Invalid version format, expected semantic version but got `${' + decodedVersion + '}`',
+    );
+  }
+
+  return decodedVersion;
+}
+
+function redirectToURL(redirectTo: URL): void {
+  const url = new URL(redirectTo.href);
+  const pathParts = url.pathname.split('/').filter(Boolean);
+  const requestedVersion = isSemanticVersion(pathParts?.[0] ?? '', 'v') ? pathParts[0] : undefined;
+
+  if (!requestedVersion) {
+    window.location.href = redirectTo.href;
+
+    return;
+  }
+
+  // Since the asset canister does not support nested fallbacks for index.html yet, we need to redirect
+  // to the exact path of the requested version so that it's index.html can be served correctly by the asset canister.
+  //
+  // The hook after the user logs in will redirect the user back to the requested path based on the session storage.
+  //
+  // Moreover, the asset canister behaves differently then response verification v2, for the later it
+  // creates a certificate where  it will lookup if an index.html exists in nested paths recursively, which means
+  // that for SPA's it would properly lookup and add the correct witness to the index.html of a path like /v1.0.0
+  // even if it were to load a dynamic route like /v1.0.0/nested-route.
+
+  // However, the asset canister if it does not have an exact path, it will always fallback to the root path for
+  // the index.html which for that example it would not work for us, since it would load the
+  // latest html for a versioned UI.
+  const unversionedUrl = new URL(window.location.href);
+  unversionedUrl.pathname = '/' + pathParts.slice(1).join('/');
+  window.sessionStorage.setItem(
+    redirectToKey,
+    unversionedUrl.pathname + unversionedUrl.search + unversionedUrl.hash,
+  );
+
+  const exactPathUrl = new URL(window.location.href);
+  exactPathUrl.pathname = '/' + requestedVersion + '/';
+
+  window.location.href = exactPathUrl.href;
+}
+
+/**
+ * Fetch the compatibility file for the UI, if a version is provided, then fetch the versioned file.
+ */
+async function fetchCompatFile(versionPath?: string): Promise<{
```

Comment:
use the type in [with-compatibility-file.plugin.ts](https://github.com/dfinity/orbit/pull/263/files#diff-5c73c14e51f334fa4113a543c72bf3172a46883b2423db5c1a86d4b2e1010e04)?

Reply:
makes sense, extracted and added to `build/types/compat.types.ts`